File: crates/cmd/onlytrades_trader/bots/mean_reversion.rs
use anyhow::Context;
use kit_ctrader_socket::*;

use crate::run_strategy::Ctx;
use crate::run_strategy::StrategyPosition;
use crate::run_strategy::StrategyState;
use crate::utils::ema::ExponentialMovingAverageExt;

const BB_LENGTH: usize = 20;
const BB_MULT: f64 = 2.0;
const RSI_LENGTH: usize = 14;
const RSI_OVERSOLD: f64 = 30.0;
const RSI_OVERBOUGHT: f64 = 70.0;
const STOP_LOSS_PCT: f64 = 0.025; // 2.5%

fn bollinger_bands(
  prices: &[f64],
  length: usize,
  mult: f64,
) -> Option<(f64, f64, f64)> {
  if prices.len() < length {
    return None;
  }
  let window = &prices[prices.len() - length..];
  let basis = window.iter().sum::<f64>() / length as f64;
  let variance = window.iter().map(|p| (p - basis).powi(2)).sum::<f64>() / length as f64;
  let std_dev = variance.sqrt();
  Some((basis, basis + mult * std_dev, basis - mult * std_dev))
}

pub async fn strategy(ctx: Ctx) -> anyhow::Result<()> {
  let prices: Vec<f64> = ctx
    .series
    .iter()
    .map(|bar| bar.close_price() as f64)
    .collect();

  if prices.len() < BB_LENGTH.max(RSI_LENGTH + 1) {
    // Not enough data yet — skip this tick
    return Ok(());
  }

  let current_close = *prices.last().context("Empty price series")?;

  // BOLLINGER BANDS
  let (bb_basis, bb_upper, bb_lower) =
    bollinger_bands(&prices, BB_LENGTH, BB_MULT).context("BB calculation failed")?;

  // RSI
  let rsi_values = ta::indicators::RelativeStrengthIndex::new(RSI_LENGTH)?.calculate(&ctx.series);
  let rsi_val = rsi_values.last().context("No RSI value")?;

  let mut state = ctx.state.lock().await;

  let long_stop_price = state.long_stop_price; // Option<f64>
  let short_stop_price = state.short_stop_price; // Option<f64>

  let long_condition = current_close < bb_lower && *rsi_val < RSI_OVERSOLD;
  let short_condition = current_close > bb_upper && *rsi_val > RSI_OVERBOUGHT;

  // ENTRY CONDITIONS
  if long_condition && state.position == StrategyPosition::Flat {
    println!(
      "📈 LONG ENTRY — close {:.0} < BB lower {:.0}, RSI {:.1}",
      current_close, bb_lower, rsi_val
    );

    // Convert stop loss % to relative points (broker units are 1/100000 of price)
    let sl_relative = (current_close * STOP_LOSS_PCT * 100_000.0) as i64;

    ctx
      .new_order(NewOrderReq {
        ctid_trader_account_id: ctx.account_id,
        symbol_id: ctx.symbol.symbol_id,
        order_type: OrderType::Market,
        trade_side: TradeSide::Buy,
        relative_stop_loss: Some(sl_relative), // broker-side backstop only
        volume: ctx.volume,
        ..NewOrderReq::default()
      })
      .await?;

    state.long_stop_price = Some(current_close * (1.0 - STOP_LOSS_PCT));
    state.position = StrategyPosition::Long;
  } else if short_condition && state.position == StrategyPosition::Flat {
    println!(
      "📉 SHORT ENTRY — close {:.0} > BB upper {:.0}, RSI {:.1}",
      current_close, bb_upper, rsi_val
    );

    let sl_relative = (current_close * STOP_LOSS_PCT * 100_000.0) as i64;

    ctx
      .new_order(NewOrderReq {
        ctid_trader_account_id: ctx.account_id,
        symbol_id: ctx.symbol.symbol_id,
        order_type: OrderType::Market,
        trade_side: TradeSide::Sell,
        relative_stop_loss: Some(sl_relative), // broker-side backstop only
        volume: ctx.volume,
        ..NewOrderReq::default()
      })
      .await?;
    state.position = StrategyPosition::Short;
    state.short_stop_price = Some(current_close * (1.0 + STOP_LOSS_PCT));
  }

  // EXIT CONDITIONS
  if state.position == StrategyPosition::Long {
    let hit_take_profit = current_close >= bb_basis;
    let hit_stop_loss = long_stop_price.map_or(false, |sl| current_close <= sl);

    if hit_take_profit {
      println!("✅ LONG TP — price reverted to mean ({:.0})", bb_basis);
      ctx.close_all_positions().await?;
      state.long_stop_price = None;
      state.position = StrategyPosition::Flat;
    } else if hit_stop_loss {
      println!(
        "🛑 LONG SL — stop loss hit at {:.0}",
        long_stop_price.unwrap()
      );
      ctx.close_all_positions().await?;
      state.long_stop_price = None;
      state.position = StrategyPosition::Flat;
    }
  }

  if state.position == StrategyPosition::Short {
    let hit_take_profit = current_close <= bb_basis;
    let hit_stop_loss = short_stop_price.map_or(false, |sl| current_close >= sl);

    if hit_take_profit {
      println!("✅ SHORT TP — price reverted to mean ({:.0})", bb_basis);
      ctx.close_all_positions().await?;
      state.short_stop_price = None;
      state.position = StrategyPosition::Flat;
    } else if hit_stop_loss {
      println!(
        "🛑 SHORT SL — stop loss hit at {:.0}",
        short_stop_price.unwrap()
      );
      ctx.close_all_positions().await?;
      state.short_stop_price = None;
      state.position = StrategyPosition::Flat;
    }
  }

  Ok(())
}

File: example_bots/mean_reversion.ts
// Bollinger Bands + RSI Mean Reversion Strategy
// Enters long when price < BB lower & RSI oversold
// Enters short when price > BB upper & RSI overbought
// Exits when price reverts to BB middle (TP) or hits stop loss

const BB_LENGTH = 20;
const BB_MULT = 2.0;
const RSI_LENGTH = 14;
const RSI_OVERSOLD = 30;
const RSI_OVERBOUGHT = 70;
const STOP_LOSS_PCT = 0.025; // 2.5%
const VOLUME = 100;

class State {
  position: "flat" | "long" | "short" = "flat";
  long_stop_price: number | undefined = undefined;
  short_stop_price: number | undefined = undefined;
}

// Runs on every candle close
export default async function handler(ctx: Context<State>) {
  const prices = ctx.series.map((bar) => bar.close);

  if (prices.length < Math.max(BB_LENGTH, RSI_LENGTH + 1)) {
    // Not enough data yet — skip this tick
    return;
  }

  const currentClose = prices[prices.length - 1];

  // Bollinger Bands
  const bb = bollingerBands(prices, BB_LENGTH, BB_MULT);
  if (!bb) return;
  const { basis, upper, lower } = bb;

  // RSI
  const rsiValues = RelativeStrengthIndex.calculate(RSI_LENGTH, ctx.series);
  const rsiValue = rsiValues[rsiValues.length - 1];

  // ENTRY CONDITIONS
  if (currentClose < lower && rsiValue < RSI_OVERSOLD && ctx.state.position === "flat") {
    console.log(
      `[${ctx.name}] 📈 LONG ENTRY — close ${currentClose.toFixed(0)} < BB lower ${lower.toFixed(0)}, RSI ${rsiValue.toFixed(1)}`,
    );

    const relativeStopLoss = currentClose * STOP_LOSS_PCT;

    await ctx.new_order({
      ctid_trader_account_id: ctx.account_id,
      symbol_id: ctx.symbol.symbol_id,
      order_type: OrderType.MARKET,
      trade_side: TradeSide.BUY,
      relative_stop_loss: relativeStopLoss,
      volume: VOLUME,
    });

    ctx.state.long_stop_price = currentClose * (1 - STOP_LOSS_PCT);
    ctx.state.position = "long";
  } else if (
    currentClose > upper &&
    rsiValue > RSI_OVERBOUGHT &&
    ctx.state.position === "flat"
  ) {
    console.log(
      `[${ctx.name}] 📉 SHORT ENTRY — close ${currentClose.toFixed(0)} > BB upper ${upper.toFixed(0)}, RSI ${rsiValue.toFixed(1)}`,
    );

    const relativeStopLoss = currentClose * STOP_LOSS_PCT;

    await ctx.new_order({
      ctid_trader_account_id: ctx.account_id,
      symbol_id: ctx.symbol.symbol_id,
      order_type: OrderType.MARKET,
      trade_side: TradeSide.SELL,
      relative_stop_loss: relativeStopLoss,
      volume: VOLUME,
    });

    ctx.state.short_stop_price = currentClose * (1 + STOP_LOSS_PCT);
    ctx.state.position = "short";
  }

  // EXIT CONDITIONS — LONG
  if (ctx.state.position === "long") {
    const hitTakeProfit = currentClose >= basis;
    const hitStopLoss =
      ctx.state.long_stop_price !== undefined && currentClose <= ctx.state.long_stop_price;

    if (hitTakeProfit) {
      console.log(
        `[${ctx.name}] ✅ LONG TP — price reverted to mean (${basis.toFixed(0)})`,
      );
      await ctx.close_all_positions();
      ctx.state.long_stop_price = undefined;
      ctx.state.position = "flat";
    } else if (hitStopLoss) {
      console.log(
        `[${ctx.name}] 🛑 LONG SL — stop loss hit at ${ctx.state.long_stop_price!.toFixed(0)}`,
      );
      await ctx.close_all_positions();
      ctx.state.long_stop_price = undefined;
      ctx.state.position = "flat";
    }
  }

  // EXIT CONDITIONS — SHORT
  if (ctx.state.position === "short") {
    const hitTakeProfit = currentClose <= basis;
    const hitStopLoss =
      ctx.state.short_stop_price !== undefined && currentClose >= ctx.state.short_stop_price;

    if (hitTakeProfit) {
      console.log(
        `[${ctx.name}] ✅ SHORT TP — price reverted to mean (${basis.toFixed(0)})`,
      );
      await ctx.close_all_positions();
      ctx.state.short_stop_price = undefined;
      ctx.state.position = "flat";
    } else if (hitStopLoss) {
      console.log(
        `[${ctx.name}] 🛑 SHORT SL — stop loss hit at ${ctx.state.short_stop_price!.toFixed(0)}`,
      );
      await ctx.close_all_positions();
      ctx.state.short_stop_price = undefined;
      ctx.state.position = "flat";
    }
  }
}

function bollingerBands(
  prices: number[],
  length: number,
  mult: number,
): { basis: number; upper: number; lower: number } | undefined {
  if (prices.length < length) return undefined;

  const window = prices.slice(prices.length - length);
  const basis = window.reduce((a, b) => a + b, 0) / length;
  const variance =
    window.reduce((acc, p) => acc + (p - basis) ** 2, 0) / length;
  const stdDev = Math.sqrt(variance);

  return { basis, upper: basis + mult * stdDev, lower: basis - mult * stdDev };
}
